pandas/core/dtypes/base.py

```diff
@@ -422,28 +427,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
 
         self.dtypes.append(dtype)
 
-    def find(self, dtype: type[ExtensionDtype] | str) -> type[ExtensionDtype] | None:
+    @overload
+    def find(self, dtype: type[ExtensionDtypeT]) -> ExtensionDtypeT:
+        ...
+
+    @overload
+    def find(self, dtype: ExtensionDtypeT) -> ExtensionDtypeT:
+        ...
+
+    @overload
+    def find(
+        self, dtype: NpDtype | type_t[str | float | int | complex | bool | object] | str
+    ) -> ExtensionDtype | None:
+        ...
+
+    def find(
+        self,
+        dtype: type[ExtensionDtype]
+        | ExtensionDtype
+        | NpDtype
+        | type_t[str | float | int | complex | bool | object],
+    ) -> type[ExtensionDtype] | ExtensionDtype | None:
         """
         Parameters
         ----------
-        dtype : Type[ExtensionDtype] or str
+        dtype : ExtensionDtype class or instance or str or numpy dtype or python type
 
         Returns
         -------
         return the first matching dtype, otherwise return None
         """
         if not isinstance(dtype, str):
-            dtype_type = dtype
+            dtype_type: type[ExtensionDtype] | type
             if not isinstance(dtype, type):
                 dtype_type = type(dtype)
+            else:
+                dtype_type = dtype
             if issubclass(dtype_type, ExtensionDtype):
-                return dtype
+                # cast needed here as mypy doesn't know we have figured
+                # out it is an ExtensionDtype
+                return cast(ExtensionDtype, dtype)
```

Comment:
`dtype` could also be `type[ExtensionDtype]` ?

Reply:
Not at this point in the logic

Reply:
using code sample from https://github.com/pandas-dev/pandas/pull/41203#discussion_r663202111

```
>>> typ
<class 'pandas.core.arrays.integer.Int64Dtype'>
>>> pd.core.dtypes.base.Registry().find(typ)
> /home/simon/pandas/pandas/core/dtypes/base.py(460)find()
-> return cast(ExtensionDtype, dtype)
(Pdb) dtype
<class 'pandas.core.arrays.integer.Int64Dtype'>
(Pdb) type(dtype)
<class 'type'>
(Pdb) 
```

Reply:
OK, I see.  Fixed in a new commit.